File: app/src/main/java/com/major/qr/ui/HomeFragment.java
package com.major.qr.ui;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.major.qr.adapters.QrDisplayAdapter;
import com.major.qr.databinding.FragmentHomeBinding;
import com.major.qr.models.Qr;
import com.major.qr.viewmodels.QrLinkViewModel;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;

public class HomeFragment extends Fragment {
    public final static String NOT_ACCESSED_YET = "Not Accessed Yet";
    private static final String TAG = HomeFragment.class.getSimpleName();
    QrLinkViewModel viewModel;
    FragmentHomeBinding binding;
    QrDisplayAdapter qrDisplayAdapter;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        binding = FragmentHomeBinding.inflate(getLayoutInflater());
        viewModel = new ViewModelProvider(requireActivity()).get(QrLinkViewModel.class);

        viewModel.getQrLink().observe(requireActivity(), jsonArray -> {
            ArrayList<Qr> list = new ArrayList<>();
            for (int i = 0; i < jsonArray.length(); ++i) {
                try {
                    JSONObject object = jsonArray.getJSONObject(i);
                    Qr qr = new Qr();
                    qr.setQrId(object.get("id").toString());
                    qr.setSessionName(object.getString("sessionName"));
                    qr.setToken(object.getString("token"));
                    if (object.has("lastSeen")) qr.setLastSeen(object.getString("lastSeen"));
                    else qr.setLastSeen(NOT_ACCESSED_YET);
                    qr.setSessionValidTime(object.getString("sessionValidTime"));
                    list.add(qr);
                } catch (JSONException e) {
                    throw new RuntimeException(e);
                }
            }
            qrDisplayAdapter = new QrDisplayAdapter(getContext(), list, viewModel);
            binding.recyclerView.setAdapter(qrDisplayAdapter);
            binding.recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
            binding.progressCircular.setVisibility(View.GONE);
            if (list.size() == 0) binding.emptyView.setVisibility(View.VISIBLE);
        });

        return binding.getRoot();
    }
}

File: app/src/main/java/com/major/qr/ui/HomeFragment.kt
package com.major.qr.ui

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import com.major.qr.adapters.QrDisplayAdapter
import com.major.qr.databinding.FragmentHomeBinding
import com.major.qr.models.Qr
import com.major.qr.viewmodels.QrLinkViewModel
import org.json.JSONArray
import org.json.JSONException

class HomeFragment : Fragment() {
    var viewModel: QrLinkViewModel? = null
    var binding: FragmentHomeBinding? = null
    var qrDisplayAdapter: QrDisplayAdapter? = null
    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentHomeBinding.inflate(layoutInflater)
        viewModel = ViewModelProvider(requireActivity()).get(QrLinkViewModel::class.java)
        viewModel!!.qrLink.observe(requireActivity(), Observer { jsonArray: JSONArray? ->
            val list = ArrayList<Qr>()
            if (jsonArray != null) {
                for (i in 0 until jsonArray.length()) {
                    try {
                        val `object` = jsonArray.getJSONObject(i)
                        val qr = Qr()
                        qr.qrId = `object`["id"].toString()
                        qr.sessionName = `object`.getString("sessionName")
                        qr.token = `object`.getString("token")
                        if (`object`.has("lastSeen")) qr.lastSeen =
                            `object`.getString("lastSeen") else qr.lastSeen = NOT_ACCESSED_YET
                        qr.sessionValidTime = `object`.getString("sessionValidTime")
                        list.add(qr)
                    } catch (e: JSONException) {
                        throw RuntimeException(e)
                    }
                }
            }
            qrDisplayAdapter = QrDisplayAdapter(requireContext(), list, viewModel!!)
            binding!!.recyclerView.adapter = qrDisplayAdapter
            binding!!.recyclerView.layoutManager = LinearLayoutManager(context)
            binding!!.progressCircular.visibility = View.GONE
            if (list.size == 0) binding!!.emptyView.visibility = View.VISIBLE
        })
        return binding!!.root
    }

    companion object {
        const val NOT_ACCESSED_YET = "Not Accessed Yet"
        private val TAG = HomeFragment::class.java.simpleName
    }
}